monitoring: update vision, revamp pillars (#2000)
sourcegraph/about: handbook/engineering/observability/monitoring_pillars.md

#### Explanation

Using the Grafana WYSIWYG editor, or otherwise operating outside the constraints of [the monitoring generator](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/monitoring) makes it:

- Much harder to integrate with our [monitoring architecture](./monitoring_architecture.md): generated dashboard integration, generated documentation, alerting support, and more.
- Extremely easy to introduce dashboards that have inconsistency errors through mistakes that are easy to make (e.g. not using UTC time on dashboards, the preset timeframe being inconsistent, it not self-describing to admins what it does, etc.)
- Difficult for others to modify and review without introducing regressions (e.g. due to Grafana schema version changes, which happen frequently, producing large diffs)

#### What you should do

You might be thinking:

> But writing code for dashboards is quite painful, let alone a bunch of generation code that I now need to understand!

> In truth, we are only using Go as a nice type-checked declarative syntax, which provides auto-completion, documentation on hover, and more for free. It's simpler to use than a configuration file, and the constraints we impose in monitoring means that we do not allow doing any fanciful dashboards intentionally -- which means you are basically _just_ writing a Prometheus query, and barely any layout or styling.

Author: This is definitely better wording, I forgot to update this from the old wording - thanks!

- [**Dashboards should only be created with the monitoring generator**](#dashboards-should-only-be-created-with-the-monitoring-generator)
- [**Dashboards should describe a single service**](#dashboards-should-describe-a-single-service)
- [**Graphs should have associated alerts**](#graphs-should-have-associated-alerts)
- [**Graphs should have less than 5 cardinality**](#graphs-should-have-less-than-5-cardinality)
- [**Less useful graphs should be hidden by default**](#less-useful-graphs-should-be-hidden-by-default)

Before trying to circumvent the guidelines enforced by our monitoring tooling, please keep in mind that these pillars are defined to help us achieve our [goals](#long-term-vision) and alleviate [pain points we have encountered in the past](#history). Each of the above pillars are documented with detailed explanations, recommendations for what you can do, and exceptions where applicable.

To make significant changes to the monitoring tooling, please reach out to the [Distribution team](../distribution/index.md) or open a PR to this page!

### Dashboards should only be created with the monitoring generator

> Creating dashboards outside of the monitoring generator, such as with the Grafana WYSIWYG editor, is discouraged. All metrics should be presented in the most plain, mundane, non-fanciful way that every site admin can understand.

Reviewer: Also disagree with this. The terms "mundane" or "non-fanciful" are completely subjective. Using whatever Grafana features to improve how you understand and navigate your graphs should be encouraged, not the other way around.

I want to try and expose the Grafana SDK in a much more direct manner in the generator.

Reviewer 2: Maybe it can be rephrased, but I do agree that we need to provide customers with easy to understand the health of the services, keeping in mind that not all customers are as technical and almost all of them do not have our context and understand of the systems.

Author: How about this approach instead:

```suggestion
Creating dashboards outside of the monitoring generator, such as with the Grafana WYSIWYG editor, is discouraged. All metrics should be presented in a set of uniform, standard ways that every site admin can understand.
```

From what I've seen, I feel like most dashboards fall into a few categories (ones that work well currently, ones that don't mostly high-cardinality graphs, that could benefit from things like labels on the right). We can create a set of formatting options that can be shared amongst everyone trying to create particular types of graphs

> keeping in mind that not all customers are as technical and almost all of them do not have our context and understand of the systems.

Lots of colours and elaborate formatting might be nice, but I think they definitely hurt the goal of making dashboards easy to interpret at a glance - _some_ expectation of how a dashboard will look going from service to service is extremely valuable IMO, especially for the uninitiated

Reviewer: >  some expectation of how a dashboard will look going from service to service is extremely valuable IMO, especially for the uninitiated

Yes, it is, but those standards and reusable patterns should evolve over time, and that can only happen if we allow experimentation. For instance, we don't consistently show any implementation of USE or RED for [Golden Signals](https://landing.google.com/sre/sre-book/chapters/monitoring-distributed-systems/#xref_monitoring_golden-signals), I think that's where we should start.

Reviewer 2: My suggestion, a bit simpler.
```suggestion
Creating dashboards outside of the monitoring generator, such as with the Grafana WYSIWYG editor, is discouraged. All metrics should be presented in a way that every site admin can understand.
```

> Yes, it is, but those standards and reusable patterns should evolve over time, and that can only happen if we allow experimentation. For instance, we don't consistently show any implementation of USE or RED for Golden Signals, I think that's where we should start.

I dont think if this is at odds with this or prevents us from creating USE and RED metrics and dashboards for those.

Reviewer: It's a really high bar. Not all site admins will understand Sourcegraph's architecture well enough that every single panel is obvious to them. What if a site-admin doesn't know that the frontend sends git commands to git server which executes them and sends back the raw output? Does that mean we shouldn't have a graph with rate of git commands over time per git server pod? Where do we draw the line?

Author: > Does that mean we shouldn't have a graph with rate of git commands over time per git server pod

I don't think this pillar is at odds with a graph like this (you might be referring to the pillar about cardinality, where we can easily say "use your best judgement, if this information only makes sense on a per-instance level"). Using the generator simply says the _presentation_ must be minimal and standard. From the rationale for this pillar:

> [not using the generator makes it] Extremely easy to introduce dashboards that have inconsistency errors through mistakes that are easy to make (e.g. not using UTC time on dashboards, the preset timeframe being inconsistent, it not self-describing to admins what it does, etc.)

Also, big hunks of JSON is hard to review and make changes to:

> [not using the generator makes it] Difficult for others to modify and review without introducing regressions

This also goes back to my previous comment:

> From what I've seen, I feel like most dashboards fall into a few categories [...] This means improvements can be more easily shared, and make future graphs easier to create

For example, you mention that the provisioning graphs look _awful_, which is true! Anything we do to make the provisioning graphs look _less_ awful, however, can easily be applied to a wide range of other graphs (namely, anything graphed per-instance), which I think is what this pillar is getting at - the generator makes it easy to make minimal, standard, consistent graphs that makes it simpler to understand.

> Yes, it is, but those standards and reusable patterns should evolve over time, and that can only happen if we allow experimentation

We do allow experimentation, since anyone can propose changes to the generator.

Could we better support this by renaming `Internal` dashboards to `Experimental`, with the expectation that they be graduated to generator dashboards?

* Remove `internal` entirely from `sourcegraph/grafana`
* Create a workflow to provision `experimental` dashboards directly through `ConfigMap` in `sourcegraph.com`

This would mean that teams can feel more free to iterate on experimental and dot-com-specific  dashboards without worrying about how site admins interpret it

Reviewer: > This would mean that teams can feel more free to iterate on experimental and dot-com-specific dashboards without worrying about how site admins interpret it

+1, but I don't know if provisioning these through a ConfigMap is better than via the Docker image. Also, I think internal is fine. What think prevented experimentation is 1) lack of access to more Grafana features and 2) an overly restrictive set of "pillars" that felt quite immovable for others (i.e. didn't invite collaboration).

Author: > I don't know if provisioning these through a ConfigMap is better than via the Docker image

Removes the worry of shipping potentially confusing dashboards to customers, and adds more impetus to work towards integrating them into the standard generator dashboards to get them shipped

> an overly restrictive set of "pillars" that felt quite immovable for others (i.e. didn't invite collaboration).

Hopefully the new more suggestion-y wording on the pillars improve this!

Reviewer: SGTM

Author: > > some expectation of how a dashboard will look going from service to service is extremely valuable IMO, especially for the uninitiated
> 
> Yes, it is, but those standards and reusable patterns should evolve over time, and that can only happen if we allow experimentation. For instance, we don't consistently show any implementation of USE or RED for [Golden Signals](https://landing.google.com/sre/sre-book/chapters/monitoring-distributed-systems/#xref_monitoring_golden-signals), I think that's where we should start.

I think the goal that we want dashboards to be consistent is valuable, so I'm going to go with my version of this pillar, unless anyone objects: https://github.com/sourcegraph/about/pull/2000#discussion_r523055237

I will see if I can squeeze in more wording about how these pillars and our tooling are open to improvement and iteration, but most of that stuff might have to wait for some implementation to catch up: https://github.com/orgs/sourcegraph/projects/112

#### Explanation

Using the Grafana WYSIWYG editor, or otherwise operating outside the constraints of [the monitoring generator](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/monitoring) makes it:

- Extremely easy to introduce dashboards that do not comply with the constraints we want to impose.
- Extremely easy to introduce dashboards that have inconsistency errors through mistakes that are easy to make (e.g. not using UTC time on dashboards, the preset timeframe being inconsistent, it not self-describing to admins what it does, etc.)
- Difficult for others to modify and review without introducing regressions (e.g. due to Grafana schema version changes, which happen frequently, producing large diffs)
- Much harder to declare proper alerts in our regular high-level alerting architecture.

> To learn more about what you get for free by using the monitoring generator, see [TODO](#TODO).

Author: I plan on figuring out a place to write up the perks of the generator somewhere! However I'm currently thinking of doing it _not_ in the handbook but directly in the `sourcegraph/sourcegraph` repo, under `monitoring` (if anyone thinks that's a bad idea let me know!)

Reviewer: SGTM

If instead this information is communicated only on a "Frontend" dashboard, it becomes immediately clear they can check the frontend logs first, which is where they will find next steps.

#### What you should do

Consider "when the metric I am monitoring is bad, where do I want the site admin to look?" Usually, this is just the service that emits the metric itself.

In other words, you should add the metric you are monitoring to an existing service's dashboard.

This also allows our [high-level alerting metrics](https://docs.sourcegraph.com/admin/observability/metrics_guide), which are generated from the same observables that each dashboard is generated from, to describe the health of Sourcegraph per-service (instead of per-service + topic like "HTTP").

### Graphs should have associated alerts

> Adding a graph/panel that visualizes a metric, without defining an associated alert, is discouraged. This includes purely informational graphs.

Reviewer: I don't agree with this. Lots of informational graphs don't need to ping anyone with an alert, and other graphs will be new enough that we don't know what good or bad is yet.

Author: > Lots of informational graphs don't need to ping anyone with an alert

Yep! There are certainly exceptions, could use your help on this! https://github.com/sourcegraph/about/pull/2000/files#r523020140

> I don't agree with this.

Do you mind elaborating on this more in the context of the existing "Explanation" section? @slimsag wrote a lot of this and I trust he had a good reason for all this, and I was personally pretty sold by this explanation :)

Reviewer: > By saying "I do not want to define an alert" what we are actually doing is shying away from the question of "how should someone interpret what this number means?" and instead saying "I will keep this information private, just for myself"

Being able to interpreting the graph doesn't mean you want to alert on it. And when you do want to alert on it, it can take time actually monitoring the service in production to learn what useful thresholds are for alerts. Best guesses are often as good as no guess.

Reviewer 3: There are also plenty of cases that give humans useful signal as to how a system is behaving, but are too variable or inconsistent to reliably alert on. for example i definitely want to be able to see at a glance the number of healthy  requests to a given service, but i don't think you could define an alert on that metric that works for all use cases. I think separating graphs from informational vs actionable is worth thinking about, and there's a lot you can gain from informational graphs or dashboards when debugging alerts from actionable ones.

Author: > it can take time actually monitoring the service in production to learn what useful thresholds are for alerts. Best guesses are often as good as no guess.

I"m not sure I follow - doesn't performing a query for a metric without a dashboard give you a look at a how the metric behaves in production already? I think the the expectation might have been to have a service export a metric in production for a while, _then_ create a dashboard based on how it behaves to ship, by which point you might have a semi-informed guess.

> Being able to interpreting the graph doesn't mean you want to alert on it

That's true too - I think there's room for improvement here overall, @pecigonzalo and I bounced the idea of properly supporting "informational" metrics with an alternative to `PossibleSolutions` (something like `Interpreting`?) to generate docs for purely information graphs. One idea was we could expand the guideline here to include something like "either you describe how to interpret it through docs, or through alerts" - how does that sound?

Reviewer: >  I think the expectation might have been to have a service export a metric in production for a while, then create a dashboard based on how it behaves to ship, by which point you might have a semi-informed guess. 

That's not the workflow I'm used to and that I think is most productive. Metrics and dashboards evolve hand in hand. I want to iterate on a dashboard, over time, in production. Perhaps there's a way to separate the "published" and "high level summary" part of a dashboard, to be consumed mostly by site admins, from the rest?

> "either you describe how to interpret it through docs

Grafana has a description field for a panel, can't we use that?

Reviewer 2: > That's not the workflow I'm used to and that I think is most productive. Metrics and dashboards evolve hand in hand. I want to iterate on a dashboard, over time, in production. Perhaps there's a way to separate the "published" and "high level summary" part of a dashboard, to be consumed mostly by site admins, from the rest?

We have to keep in mind this is not just an internal dashboard and are part of the overall product which we ship to customers monthly and remain there for a period of time. I agree that dashboards evolve over time, but I believe its better to think about them as we think about our services, we iterate over in production by committing, PR and shipping. Dashboards are part of the site-admin experience and we should treat them, like our APIs and site-config, etc.
If we have the tooling to support it, do you think developing locally against the remote Prometheus works? If not, what are the pitfalls?

> Grafana has a description field for a panel, can't we use that?

I would prefer to have it in our documentation instead, as we are lacking a lot information on that front, many metrics I dont know how to interpret and It would be helpful to have some information.

Reviewer: > If we have the tooling to support it, do you think developing locally against the remote Prometheus works? If not, what are the pitfalls?

Not as well, since we can't share the dashboard with others as easily. Your concerns would be better addressed by having "internal" groups in each dashboard that don't have to abide by the same standards.

> I would prefer to have it in our documentation instead, as we are lacking a lot information on that front, many metrics I dont know how to interpret and It would be helpful to have some information.

I think both make sense.

Author: > Not as well, since we can't share the dashboard with others as easily.

How would you normally share it? Right now you should be able to share the dashboard JSON generated by the generator

> I would prefer to have it in our documentation instead, as we are lacking a lot information on that front, many metrics I dont know how to interpret and It would be helpful to have some information.

Link to docs in description, similar to those generated for alerts? :)

> Perhaps there's a way to separate the "published" and "high level summary" part of a dashboard, to be consumed mostly by site admins

You can do so already by creating a new group in the generator (even better now without required alerts)

Author: > Not as well, since we can't share the dashboard with others as easily.

How would you normally share it? Right now you should be able to share the dashboard JSON generated by the generator.

Author: In response to this feedback, and [the other thread](https://github.com/sourcegraph/about/pull/2000#discussion_r523050126), I have completely removed this pillar in favour of adding more emphasis on "just use the generator" to ensure documentation is in place: 8662cba269a6255dfc4a16fa7f9242dcb8d25a1a

This will require some implementation work to enforce docs on graphs without alerts. I have added it to the PR description, and will open up more specific issues when we've finalized this PR :)

Reviewer: > How would you normally share it? Right now you should be able to share the dashboard JSON generated by the generator.

Link to the live dashboard. Example: https://sourcegraph.com/-/debug/grafana/d/QvLV8J2Gk/git-server-cloud?orgId=1&from=1605082470899&to=1605083727190&var-shard=All

Reviewer 2: > Not as well, since we can't share the dashboard with others as easily. Your concerns would be better addressed by having "internal" groups in each dashboard that don't have to abide by the same standards.

Could you elaborate on this? I'm not mapping how internal dashboards address my comment regarding develop in production and the workflow used for metrics at other places in which the dashboards are not shipped to customers and released as part of the product.
For example, we dont develop it in production by just changing the live files, we instead send a PR which can be reviewed, etc. I think these dashboards have to be changed in the same way.

I do think having some internal dashboards that we can iterate over a bit faster, likely not tied to alerts but rather information that we can use to debug .com and customer environments. Or just experimental dashboards that are not "ready" yet as discussed in the other thread and as we talked about in our conversation when we said "Is there room for 2 kinds of dashboards?".

Coming back to the development experience, is the main point the inability to share the dashboard directly via a link? How are we currently doing when developing other interfaces?

Reviewer: > Coming back to the development experience, is the main point the inability to share the dashboard directly via a link? How are we currently doing when developing other interfaces?

Yeah, quick iteration and visual feedback, and sharing via a link with other teammates for more feedback, analysis, etc.

Author: > > Coming back to the development experience, is the main point the inability to share the dashboard directly via a link? How are we currently doing when developing other interfaces?
> 
> Yeah, quick iteration and visual feedback, and sharing via a link with other teammates for more feedback, analysis, etc.

I imagine for other interfaces it's just screenshots at the moment, right? Assuming that work is just as iterative, I'm not sure if it is really that uncomfortable to go without the link sharing. This seems like a reasonable workflow:

1. Port-forward prometheus
2. `git checkout ...`
3. `./dev/grafana.sh`
4. marvel at dashboards

It's not necessary to spin up everything `dev/start`, so dashboards arguably already have the best sharing/collaboration experience at the moment (maybe?)

> For example, we dont develop it in production by just changing the live files, we instead send a PR which can be reviewed, etc. I think these dashboards have to be changed in the same way.

+1, and this seems like a good reason to _not_ use the Grafana editor to make edits in the Grafana database and share a link with teammates (I think this is the only way to get the workflow Tomas envisions)

Author: > I do think it will be good to have some internal dashboards that we can iterate over a bit faster, likely not tied to alerts but rather information that we can use to debug .com and customer environments. Or just experimental dashboards that are not "ready" yet as discussed in the other thread and as we talked about in our conversation when we said "Is there room for 2 kinds of dashboards?".

Do we envision needing experimental dashboards for customers? I don't think they upgrade frequently enough for this to be a good way to iterate - if we do need this, we can send them dashboard JSON as required, and for dot-com do experimental dashboard provisioning via ConfigMap: https://github.com/sourcegraph/about/pull/2000#discussion_r523864481 

This workflow would then look like this:

* Provision experimental dashboards in via `deploy-sourcegraph-dot-com`. Iterate freely without need for review because we don't ship these to customers.
* Iterate on experimental dashboards in customer instances by sending them dashboard JSON. Iterate freely because we can just keep sending them more dashboard JSON without having to ask them to upgrade to a new, shipped Grafana image.
* Remove the current custom dashboards shipped with the Grafana image to drive eventually "promoting" experimental dashboards to generator dashboards, and to have more impetus to improve the generator to enable this if required (to avoid becoming increasingly dependent on raw dashboards)

Reviewer: SGTM

Author: Created action items to make this workflow happen in the PR description - resolving this conversation for now :) Thanks for the discussion!

See also ["the difference between Warning and Critical alerts"](https://docs.sourcegraph.com/admin/observability/alerting_custom_consumption#the-difference-between-critical-and-warning-alerts) for how we communicate this to site admins.

#### What you should do

Declare a `Warning` alert, and give it your best guess at what a threshold would look like. It's OK if it's flaky, or fires when there isn't a problem sometimes or on some Sourcegraph instances.

Over time, we will adjust the alert so it becomes more reliable and becomes the baseline number that most Sourcegraph instances observe. Once it becomes really reliable (and empirically proven to be so, based on pings we get from customer instances), it can be promoted to a `Critical` alert as well if it speaks of a critical part of Sourcegraph.

#### Exceptions

> TODO

Author: @tsenart TODO what graphs are good to _not_ include alerts on

Reviewer: There are no specific graphs. Building a dashboard is an iterative endeavour, over the course of many deployments, observation of graphs to learn what good and bad is, and subsequent tweaking, and *possibly* definition of alerts, when useful / applicable.

Author: I think @arussellsaw's description here https://github.com/sourcegraph/about/pull/2000#discussion_r523034778 provides a step towards a concrete exception :) Some guideline for "informational vs actionable" would be good to have here!

Reviewer 3: to be clear my personal take is that engineers should be as free as possible to make whichever informational dashboards as they want, but actionable ones should absolutely meet certain standards

Author: > should be as free as possible to make whichever informational dashboards as they want

I think we should properly support informational dashboards, but this particular snippet I thought is a good argument for having _some_ recommendation towards one or the other: https://github.com/sourcegraph/about/pull/2000/files#diff-c74c0b112f5eaedf8f54f56b54ebaff1294645e9d10106829776938eaec60d0aR186

> [...] it's easier for us to say:
>
>> Go to the **Searcher** dashboard, scroll down and expand the **Internal service requests** panel and send us a screenshot.
>
> Than it is for us to say:
>
>> On the **Searcher** dashboard, send us a screenshot of the whole page (I know it's really long, sorry about that, if you could just make sure to include the "Frontend request failures over 5m" panel and the [...] panels, that is all we should need)

Author: Though now that I think of it, maybe it's not really that important and teams should just make sure their dashboards are of reasonable size themselves :)

Author: Resolving, since I just removed the pillar: https://github.com/sourcegraph/about/pull/2000#discussion_r523886716

![](https://user-images.githubusercontent.com/3173176/81974154-03f4b780-95da-11ea-8988-069dbdc3ef92.png)

#### Exceptions

> Some graphs will require per-instance granularity. For example, per-container or per-Gitserver replica graphs might be rendered entirely unhelpful if we do not have the ability to observe the behaviour of each individual instance.

Author: > I honestly don't know what is a good "parameter" for this, its just a matter of how easy it is to grok as human.

I like this - I updated your suggested edit to include this, does it look good to you?

#### What you should do

If the information you are presenting is important enough to warrant being shown independently, it should be shown in its own panel. In other words, declare a separate `Query` all-together with a Prometheus selector like `{method="update_commits"}`.

If the information you are presenting is not important enough to warrant being shown independently, or if you're not sure, then aggregate it! It's far better that we have clear panels showing individual numbers and learn we need to add more for specific cases, than it is for us to have every bit of information present such that it is impossible to interpret. We want a low signal-to-noise ratio.

For _instances of services_ in particular, you might be thinking "surely that is an exception?" But no, it is not generally. It is better that an admin knows there is an issue with _a searcher_, or _a gitserver_ and that admin looks through the logs of each to determine the issue than it is to give an information overload which makes it hard to interpret individual panels at all.

> ### Less useful graphs should be hidden by default

Reviewer 2: ```suggestion
### Complex or less frequently used graphs should be hidden by default
```

Maybe this is a better way to phrase it. I dont know if they are less useful, they might be just too granular, or  otherwise consulted less frequently.

Again, I would not know how to phrase it but I think its basically the "at a glance" vs "deep dive troubleshooting" view.

Author: On second thought, what do you think about inverting this pillar?

```suggestion
### Only the most useful and simple graphs should be visible by default
```

Reviewer 2: I like the idea of inverting it, Im not set on the way to phrase it, but LGTM to get started, we can change it later if we find a better way to express it.
